### PR TITLE
[MKM] Implement Kaya, Spirits' Justice and new zone change batch event

### PR DESCRIPTION
**Pull request by DominionSpy**
#11516

### CODE REVIEW

File: Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        ability.addTarget(new TargetPermanent(StaticFilters.FILTER_CONTROLLED_CREATURE));
+
+        for (UUID playerId : game.getState().getPlayersInRange(ability.getControllerId(), game)) {
+            Player player = game.getPlayer(playerId);
+            if (player == null || player == controller) {
+                continue;
+            }
+            FilterPermanent filter = new FilterCreaturePermanent("creature that player controls");
+            filter.add(new ControllerIdPredicate(playerId));
+            ability.addTarget(new TargetPermanent(0, 1, filter));
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java:263]: Use target withChooseHint(“from “ + player.getlogname) so you can hint user about current player name

File: Mage.Tests/src/test/java/org/mage/test/cards/single/mkm/KayaSpiritsJusticeTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @author DominionSpy
+ */
+public class KayaSpiritsJusticeTest extends CardTestPlayerBase {
+
+    @Test
+    public void test_TriggeredAbility() {
+        addCard(Zone.BATTLEFIELD, playerA, "Plains", 1 + 2 + 6);
+        addCard(Zone.BATTLEFIELD, playerA, "Kaya, Spirits' Justice");
```
Comment (JayDi85) [inline Mage.Tests/src/test/java/org/mage/test/cards/single/mkm/KayaSpiritsJusticeTest.java:17]: Must add all used card rules as comments here or as card name const.

Example:
IMG_9621

File: Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        Permanent copyToPermanent = game.getPermanent(source.getFirstTarget());
+        Cards exiledCards = new CardsImpl(getTargetPointer().getTargets(game, source));
+        if (controller == null || copyToPermanent == null || exiledCards.isEmpty()) {
+            return false;
+        }
+
+        TargetCard target = new TargetCardInExile(0, 1, StaticFilters.FILTER_CARD_CREATURE, null);
+        if (!target.canChoose(source.getControllerId(), source, game)) {
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java:189]: No needs to check canChoose — chooseTarget must process it

File: Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            return false;
+        }
+
+        Card copyFromCard = game.getCard(target.getFirstTarget());
+        if (copyFromCard == null) {
+            return false;
+        }
+
+        game.addEffect(new CopyEffect(Duration.EndOfTurn, copyFromCard, copyToPermanent.getId()), source);
+        game.addEffect(new GainAbilityTargetEffect(FlyingAbility.getInstance(), Duration.EndOfTurn), source);
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java:202]: It’s part of copy, must use CopyApplier instead gain.

Example:
IMG_9622

**Comment by xenohedron** [comment]
So what is the difference between `ZONE_CHANGE_GROUP` and `ZONE_CHANGE_BATCH`?

**Comment by DominionSpy** [comment]
> So what is the difference between `ZONE_CHANGE_GROUP` and `ZONE_CHANGE_BATCH`?
> 
> 
> 
> Code looks fine now, just want to make sure documentation is clear for future devs.

`ZONE_CHANGE_BATCH` contains all zone changes that occurred in a single action, whereas `ZONE_CHANGE_GROUP` only contains changes that occurred between the same two zones.

For example, paying a Craft cost by exiling a creature permanent and a card from your graveyard would fire 1 `ZONE_CHANGE_BATCH` event but 2 `ZONE_CHANGE_GROUP` events. In this instance, Kaya's ability should only trigger once.

File: Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+        Player controller = game.getPlayer(ability.getControllerId());
+        if (controller == null) {
+            return;
+        }
+        ability.addTarget(new TargetPermanent(StaticFilters.FILTER_CONTROLLED_CREATURE));
+
+        for (UUID playerId : game.getState().getPlayersInRange(ability.getControllerId(), game)) {
+            Player player = game.getPlayer(playerId);
+            if (player == null || player == controller) {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java:273]: compare UUID with equals rather than player objects

**Comment by JayDi85** [comment]
ZONE_CHANGE_GROUP and related usages will be replaced by ZONE_CHANGE_BATCH in the future due buggy/wrong logic. Batch logic is correct.

**Review by JayDi85 — APPROVED** [review]

**Pull request merged**